File: 2024/04/test_day_04.py
import unittest
from day_04 import Solution

class TestSolution(unittest.TestCase):
    def setUp(self):
        self.solution = Solution()

    def test_find_words_aoc_input(self):
        words = [x.strip() for x in """....XXMAS.
.SAMXMS...
...S..A...
..A.A.MS.X
XMASAMX.MM
X.....XA.A
S.S.S.S.SS
.A.A.A.A.A
..M.M.M.MM
.X.X.XMASX""".splitlines()]
        result = self.solution.find_words(words, "XMAS")
        self.assertEqual(result, 18)

    def test_find_words_simple_input(self):
        words = [x.strip() for x in """SSXMAS
                 SAMXMS""".splitlines()]
        result = self.solution.find_words(words, "XMAS")
        self.assertEqual(result, 2)

    def test_find_words_simple_diagonal(self):
        words = [x.strip() for x in """XZZS
                                       ZMAZ
                                       ZMAZ
                                       XZZS""".splitlines()]
        result = self.solution.find_words(words, "XMAS")
        self.assertEqual(result, 2)

    def test_find_words_x_simple_input(self):
        words = [x.strip() for x in """M.S
                                       .A.
                                       M.S""".splitlines()]
        result = self.solution.find_words_x(words, "MAS")
        self.assertEqual(result, 1)

    def test_find_words_x_reverse_input(self):
        words = [x.strip() for x in """S.M
                                       .A.
                                       S.M""".splitlines()]
        result = self.solution.find_words_x(words, "MAS")
        self.assertEqual(result, 1)

    def test_find_words_x_mixed_input(self):
        words = [x.strip() for x in """M.M
                                       .A.
                                       S.S""".splitlines()]
        result = self.solution.find_words_x(words, "MAS")
        self.assertEqual(result, 1)

    def test_find_words_x_superfluous_input(self):
        words = [x.strip() for x in """M.M
                                       .A.
                                       M.M""".splitlines()]
        result = self.solution.find_words_x(words, "MAS")
        self.assertEqual(result, 0)

    def test_find_words_x_aoc_input(self):
        words = [x.strip() for x in """.M.S......
                                       ..A..MSMS.
                                       .M.S.MAA..
                                       ..A.ASMSM.
                                       .M.S.M....
                                       ..........
                                       S.S.S.S.S.
                                       .A.A.A.A..
                                       M.M.M.M.M.
                                       ..........""".splitlines()]
        result = self.solution.find_words_x(words, "MAS")
        self.assertEqual(result, 9)
if __name__ == '__main__':
    unittest.main()

File: 2024/04/test_day_04.jl
using Test
include("./day_04.jl")

# Unit tests
function test_solution()
    solution = Solution()

    @test find_words(solution, [String(strip(x)) for x in split("""SSXMAS
                                                                   SAMXMS""", "\n")], "XMAS") == 2
    @test find_words(solution, [String(strip(x)) for x in split("""....XXMAS.
                                                                   .SAMXMS...
                                                                   ...S..A...
                                                                   ..A.A.MS.X
                                                                   XMASAMX.MM
                                                                   X.....XA.A
                                                                   S.S.S.S.SS
                                                                   .A.A.A.A.A
                                                                   ..M.M.M.MM
                                                                   .X.X.XMASX""", "\n")], "XMAS") == 18
    @test find_words(solution, [String(strip(x)) for x in split("""XZZS
                                                                   ZMAZ
                                                                   ZMAZ
                                                                   XZZS""", "\n")], "XMAS") == 2
    @test find_words_x(solution, [String(strip(x)) for x in split("""M.S
                                                                     .A.
                                                                     M.S""", "\n")], "MAS") == 1
    @test find_words_x(solution, [String(strip(x)) for x in split("""S.M
                                                                     .A.
                                                                     S.M""", "\n")], "MAS") == 1
    @test find_words_x(solution, [String(strip(x)) for x in split("""M.M
                                                                     .A.
                                                                     S.S""", "\n")], "MAS") == 1
    @test find_words_x(solution, [String(strip(x)) for x in split("""S.M
                                                                     .A.
                                                                     M.S""", "\n")], "MAS") == 0

                                                                     
    @test find_words_x(solution, [String(strip(x)) for x in split(""".M.S......
                                                                     ..A..MSMS.
                                                                     .M.S.MAA..""", "\n")], "MAS") == 1

    @test find_words_x(solution, [String(strip(x)) for x in split("""AMASAAAAAA
                                                                     A.A..MSMS.
                                                                     .M.S.MAA..""", "\n")], "MAS") == 1

    @test find_words_x(solution, [String(strip(x)) for x in split("""AMASAAAAAA
                                                                     AAAAAMSMSA
                                                                     AMASAMAAAA
                                                                     AAAAASMSMA
                                                                     AMASAMAAAA
                                                                     AAAAAAAAAA""", "\n")], "MAS") == 5

    @test find_words_x(solution, [String(strip(x)) for x in split(""".M.S......
                                                                     ..A..MSMS.
                                                                     .M.S.MAA..
                                                                     ..A.ASMSM.
                                                                     .M.S.M....
                                                                     ..........
                                                                     S.S.S.S.S.
                                                                     .A.A.A.A..
                                                                     M.M.M.M.M.
                                                                     ..........""", "\n")], "MAS") == 9
end

test_solution()